### ISSUE TITLE
Intel oneMKL

### ISSUE THREAD

**Issue by ChufanSuki** [ChufanSuki/read-paper-and-code#145]
### What is BLAS

Basic Linear Algebra Subprograms (BLAS) is a specification that prescribes a set of low-level routines for performing common linear algebra operations such as vector addition, scalar multiplication, dot products, linear combinations, and matrix multiplication. 

There are now three levels of BLAS:

- Level 1 BLAS:

for vector operations, such as $\mathbf{y} \leftarrow \alpha \mathbf{x} + \mathbf{y}$(called "axpy", "a x plus y").

- Level 2 BLAS:

for matrix-vector operations, such as $\mathbf{y} \leftarrow \alpha A \mathbf{x} + \beta \mathbf{y}$(called "gemv")

as well as a solver for $\mathbf{x}$ in the linear equation

```math
T \mathbf{x} = \mathbf{y}
```
with $T$ being triangular.

- Level 3 BLAS:

for matrix-matrix operations, such as 

```math
C \leftarrow \alpha AB + \beta C
```

They are the de facto standard low-level routines for linear algebra libraries; The routines have bindings for both C ("CBLAS interface") and Fortran("BLAS interface").  

Although the BLAS specification is general, BLAS implementations are often optimized for speed on a particular machine, so using them can bring substantial performance benefits.

#### Implementations 

- cuBLAS(NVIDIA GPU)
- rocBLAS(AMD GPU)
- OpenBLAS(CPU)
- Intel oneAPI Math Kernel Library(Intel CPU/GPU)
- Accelerate(Apple)

### What is LAPACK

LAPACK ("**L**inear **A**lgebra **Pack**age") is a standard software library for numerical linear algebra. It provides routines for solving systems of linear equations and linear least squares, eigenvalue problems, and singular value decomposition. It also includes routines to implement the associated matrix factorizations such as LU, QR, Cholesky and Schur decomposition.

LAPACK relies on an underlying _BLAS_ implementation to provide efficient and portable computational building blocks for its routines.

#### Implementations 

- Accelerate(Apple)
- Netlib LAPACK
- OpenBLAS
- Eigen
- Intel oneAPI Math Kernel Library

#### The BLAS as the Key to Portability

The LAPACK strategy for combining efficiency with portability is to construct the software as much as possible out of calls to the BLAS (Basic Linear Algebra Subprograms); the BLAS are used as building blocks.

The efficiency of LAPACK software depends on efficient implementations of the BLAS being provided by computer vendors (or others) for their machines. Thus the BLAS form a low-level interface between LAPACK software and different machine architectures. Above this level, almost all of the LAPACK software is truly portable.

#### Numpy

When a NumPy build is invoked, BLAS and LAPACK library detection happens automatically. The build system will attempt to locate a suitable library, and try a number of known libraries in a certain order - most to least performant. A typical order is: `MKL`, `Accelerate`, `OpenBLAS`, `FlexiBLAS`, `BLIS`, plain `libblas`/`liblapack`.

#### PyTorch

One can set `BLAS` option when building PyTorch`set BLAS=OpenBLAS && python setup.py install` It can be

- MKL
- Eigen
- ATLAS
- FlexiBLAS
- OpenBLAS

If set then the build will fail if the requested BLAS is not found, otherwise the BLAS will be chosen based on what is found on your system.

> FlexiBLAS is a wrapper library that enables the exchange of the [BLAS](https://en.wikipedia.org/wiki/BLAS) (Basic Linear Algebra System) and [LAPACK](https://en.wikipedia.org/wiki/LAPACK) (Linear Algebra PACKage) implementation used in an executable without recompiling or re-linking it.

> Automatically Tuned Linear Algebra Software (ATLAS) is a software library for linear algebra. It provides a mature open source implementation of BLAS APIs for C and FORTRAN 77.
ATLAS is often recommended as a way to automatically generate an optimized BLAS library. While its performance often trails that of specialized libraries written for one specific hardware platform, it is often the first or even only optimized BLAS implementation available on new systems and is a large improvement over the generic BLAS available at Netlib. For this reason, ATLAS is sometimes used as a performance baseline for comparison with other products.

| Operate| Intel CPU Library Support| Nvidia GPU Library Support|
|--------|--------|--------|
| Convolution| dnnl::convolution_forward / dnnl::convolution_backward_weights / dnnl::convolution_backward_data | cudnnConvolutionForward / cudnnConvolutionBackwardFilter / cudnnConvolutionBackwardData |
| Dot| cblas_sgemm / cblas_sgemm_batch| cublasSgemm / cublasSgemmBatched / cublasGemmEx|
| BatchNormalzation| dnnl::batch_normalization_forward / dnnl::batch_normalization_backward | cudnnBatchNormalizationForwardInference / cudnnBatchNormalizationForwardTraining / cudnnBatchNormalizationBackward |

**Comment by ChufanSuki** [comment]
### Intel oneAPI MKL System Requirements

See: https://www.intel.com/content/www/us/en/developer/articles/system-requirements/oneapi-math-kernel-library-system-requirements.html

#### Hardware Requirements

Intel CPU and Intel GPU

#### Software Requirements 

##### Operating System

Windows and Linux

##### Compiler 

Intel oneAPI DPC++ and GNU GCC

**Comment by ChufanSuki** [comment]
### Intel oneAPI MKL's Routines and Functions

See https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-c

- BLAS Routines
- Sparse BLAS Routines
- Sparse QR
- LAPACK Routines
- Sparse Solver Routines
- Extended Eigensolver Routines
- VM Functions
- Statistical Functions
- Fourier Transform Functions
- Partial Differential Equations Support
- Support Functions

#### Multiplying Matrices Using dgemm

`dgemm` is the most widely used  routines in oneMKL
- 'd' refers to data type: 
  - 's': single-precision real
  - 'd': double-precision real
  - 'c': single-precision complex 
  - 'z': double-precision complex 
- 'gemm' means **ge**neral **m**atrix **m**ultiply

```math
C \leftarrow \alpha AB + \beta C
```

```c
/*******************************************************************************
*   This example computes real matrix C=alpha*A*B+beta*C using Intel(R) MKL
*   function dgemm, where A, B, and C are matrices and alpha and beta are
*   scalars in double precision.
*
*   In this simple example, practices such as memory management, data alignment,
*   and I/O that are necessary for good programming style and high MKL
*   performance are omitted to improve readability.
********************************************************************************/

#define min(x,y) (((x) < (y)) ? (x) : (y))

#include <stdio.h>
#include <stdlib.h>
#include "mkl.h"

int main()
{
    double *A, *B, *C;
    int m, n, p, i, j;
    double alpha, beta;

    printf ("\n This example computes real matrix C=alpha*A*B+beta*C using \n"
            " Intel(R) MKL function dgemm, where A, B, and  C are matrices and \n"
            " alpha and beta are double precision scalars\n\n");

    m = 1024, p = 1024, n = 1024;
    printf (" Initializing data for matrix multiplication C=A*B for matrix \n"
            " A(%ix%i) and matrix B(%ix%i)\n\n", m, p, p, n);
    alpha = 1.0; beta = 0.0;

    printf (" Allocating memory for matrices aligned on 64-byte boundary for better \n"
            " performance \n\n");
    A = (double *)mkl_malloc( m*p*sizeof( double ), 64 );
    B = (double *)mkl_malloc( p*n*sizeof( double ), 64 );
    C = (double *)mkl_malloc( m*n*sizeof( double ), 64 );
    if (A == NULL || B == NULL || C == NULL) {
      printf( "\n ERROR: Can't allocate memory for matrices. Aborting... \n\n");
      mkl_free(A);
      mkl_free(B);
      mkl_free(C);
      return 1;
    }

    printf (" Intializing matrix data \n\n");
    for (i = 0; i < (m*p); i++) {
        A[i] = (double)(i+1);
    }

    for (i = 0; i < (p*n); i++) {
        B[i] = (double)(-i-1);
    }

    for (i = 0; i < (m*n); i++) {
        C[i] = 0.0;
    }

    printf (" Computing matrix product using Intel(R) MKL dgemm function via CBLAS interface \n\n");
    cblas_dgemm(CblasRowMajor, CblasNoTrans, CblasNoTrans,
                m, n, p, alpha, A, p, B, n, beta, C, n);
    printf ("\n Computations completed.\n\n");

    printf (" Top left corner of matrix A: \n");
    for (i=0; i<min(m,6); i++) {
      for (j=0; j<min(p,6); j++) {
        printf ("%12.0f", A[j+i*p]);
      }
      printf ("\n");
    }

    printf ("\n Top left corner of matrix B: \n");
    for (i=0; i<min(p,6); i++) {
      for (j=0; j<min(n,6); j++) {
        printf ("%12.0f", B[j+i*n]);
      }
      printf ("\n");
    }

    printf ("\n Top left corner of matrix C: \n");
    for (i=0; i<min(m,6); i++) {
      for (j=0; j<min(n,6); j++) {
        printf ("%12.5G", C[j+i*n]);
      }
      printf ("\n");
    }

    printf ("\n Deallocating memory \n\n");
    mkl_free(A);
    mkl_free(B);
    mkl_free(C);

    printf (" Example completed. \n\n");
    return 0;
}
```

##### oneMKL Initialization on CPU

When a user first invokes any oneMKL functions, there is some initialization routines:

- setting routine `xerbla` that acts as an error handler 
- setting oneMKL verbose mode with either the environment variable `MKL_VERBOSE` or function `mkl_verbose(int mode)`
-

**Comment by ChufanSuki** [comment]
### BLAS and Sparse BLAS Routines

Naming convention: <character> <name> <mod> [_64]

- <character>: data type
- <name>: In BLAS level 1, indicates the operation type. In BLAS level 2 and 3, indicates the matrix argument type
- <mod>: additional details of the operation
- _64 suffix: support large data arrays in the LP64 interface library on 64-bit platforms

CBLAS, the C interface to the Basic Linear Algebra Subprograms (BLAS), provides a C language interface to BLAS routines. In CBLAS, the Fortran routine names are prefixed with `cblas_` (for example, `dasum` becomes `cblas_dasum`).

**Comment by ChufanSuki** [comment]
### PyTorch

BlasKernel.cpp

- `sdot_`: `cblas_sdot`
- `cdotu_`: `cblas_cdotu_sub`
- `zdotu_`: `cblas_zdotu_sub`
- `cdotc_`: `cblas_cdotc_sub`
- `zdotc_`: `cblas_zdotc_sub`

CPUBlas.cpp

- `gemm`: `cblas_dgemm`, `cblas_sgemm`, `cblas_zgemm`, `cblas_cgemm`, `cblas_gemm_i64_i64acc`
- `axpy`: `cblas_daxpy`, `cblas_saxpy`, `cblas_zaxpy`, `cblas_caxpy`
- `copy`: `cblas_dcopy`, `cblas_scopy`, `cblas_zcopy`, `cblas_ccopy`

...

**Comment by ChufanSuki** [comment]
### Benchmark

- https://github.com/andre-wojtowicz/blas-benchmarks

- Intel CPU Matrix Multiplication

![image](https://github.com/ChufanSuki/read-paper-and-code/assets/50797312/85ab37e5-5573-4710-a529-8d6323ceab80)

- AMD CPU Matrix Multiplication

| BLAS| AVg Runtime|
|--------|--------|
| openblas| 1.003829|
| mkl2024| 1.128423| 
| mkl2020.1_fakeintel| 1.144065|
